src/main/java/org/apache/commons/codec/binary/BaseNCodec.java

```diff
@@ -569,7 +570,11 @@ public String encodeToString(final byte[] pArray) {
      */
     protected byte[] ensureBufferSize(final int size, final Context context){
         if (context.buffer == null) {
-            context.buffer = new byte[Math.max(size, getDefaultBufferSize())];
+            int bufferSize = Math.max(size, getDefaultBufferSize());
+            if (context.bufferCache == null || context.bufferCache.length != bufferSize) {
```

Comment:
Base16 seems to have variable sizes. I don't understand the inner workings well enough currently. Will have to dig deeper. Wouldn't mind if you took over, though. :)

Reply:
Ideally we wouldn't even need the cache, but the buffer is set to null to signal the calling code. I don't feel competent enough to refactor that though.

Reply:
I see now that encode and decode are called in BaseNCodecInputStream.read with the number of bytes read into the buffer. This can be used to optimise allocation as is done in Base16 codec. But given that the number of bytes will not be more than 8196 then this may suffice to better recycle the buffer:
```java
if (context.bufferCache == null || context.bufferCache.length < bufferSize) {
```
This will work as the buffer size is not required to be exactly the size requested, only that it must be equal to or larger than the size requested.

Use of the buffer as null for a signal predates my involvement with these classes. The use of buffer compared to null or not null is only done in a few places; all those places also use the `readPos` marker in the same method or directly after as they are all inquiring about whether there is anything to read. The same functionality may be achieved using `pos > readPos` as the signal. I updated BaseNCodec using (I left the old code commented out for reference):
```java
    int available(final Context context) {  // package protected for access from I/O streams
        return hasData(context) ? context.pos - context.readPos : 0;
        //return context.buffer != null ? context.pos - context.readPos : 0;
    }

    boolean hasData(final Context context) {  // package protected for access from I/O streams
        return context.pos > context.readPos;
        //return context.buffer != null;
    }

    int readResults(final byte[] b, final int bPos, final int bAvail, final Context context) {
        if (hasData(context)) {
        //if (context.buffer != null) {
            final int len = Math.min(available(context), bAvail);
            System.arraycopy(context.buffer, context.readPos, b, bPos, len);
            context.readPos += len;
            if (context.readPos >= context.pos) {
                // All data read.
                // Reset markers so hasData() will return false, and this method can return -1,
                // but do not set buffer to null to allow reuse.
                context.pos = context.readPos = 0;
            //    context.buffer = null; // so hasData() will return false, and this method can return -1
            }
            return len;
        }
        return context.eof ? EOF : 0;
    }
```

This passes all the unit tests and does not require the buffer be set to null.

You could update the PR to use this method and then remove the bufferCache. The buffer itself will then be recycled.

Reply:
Thanks! I'll rework the PR.